### DISCUSSION TITLE
next/image warns when one of multiple srcset urls matches the src

### DISCUSSION THREAD

**Discussion by ShuPink** [vercel/next.js#37102] [Help] [answered]
Version: next@12.1.6

__Summary__

- I am trying to use next/image's custom loader with cloudinary with `layout="fill"`.  
- I'm getting an error when one (or more) urls generated by the loader in the srcSet matches the src. 
- I'm not sure how to resolve the warning with my set up.

__Context__

I have my own transforms (in cloudinary). 
When a user uploads an image it will eagerly generate smaller images (amongst other things). 
I'm using using different params from the cloudinary loader, so I am using the custom loader.

__Example__
- the original image size which is the largest at 3440px.
- generated preset resizes e.g.: 234px, 750px,1000px

```
// next config
  images: {
    domains: ['localhost', 'res.cloudinary.com'],
    imageSizes: [234],
    deviceSizes: [600, 1000, 3440],
    path: 'https://res.cloudinary.com/<cloud-name>/',
    loader: 'custom',
  },
```

My loader spits out the srcSet which is as desired, where the 3440w image is the same as the src image.  

```
https://res.cloudinary.com/<name>/image/upload/w_234/v1/test/pic.jpg 234w, 
https://res.cloudinary.com/<name>/image/upload/w_750/v1/test/pic.jpg 600w, 
https://res.cloudinary.com/<name>/image/upload/w_1000/v1/test/pic.jpg 1000w, 
https://res.cloudinary.com/<name>/image/upload/v1/test/pic.jpg 3440w <-- same as src
```

next/image then gives me the warning because the 3440w url in the srcSet is the same as the src.

```
Image with src "${src}" has a "loader" property that does not implement width. Please implement it or use the "unoptimized" property instead.`
```

I don't want to use unoptimized as it removes the srcset.
Is it wrong to use a url which is the same as the src in the srcSet?  
Am I just going about this the wrong way?

**Comment by icyJoseph** [comment, marked as answer]
I may be wrong here, but it feels like the check is just there to ensure you don't unintentionally make a mistake. Not sure if there's any actual technical problems arising from it.

You could just add the width to the URL, I guess? or name it `/image/upload/default/v1/test/pic.jpg` in your loader. 

https://github.com/vercel/next.js/blob/canary/packages/next/client/image.tsx#L639

> This error occurred because the provided loader() function did not use width in the returned URL string. This means that the image will likely not be resized and therefore degrade performance.

I guess from their point of view, you couldn't really optimise the image, if you don't know what width to use. Which is a fair warning.

**Reply by ShuPink** [reply]
Hiya, thanks for taking the time to respond
I'm not sure I can name it like `/image/upload/default/v1/test/pic.jpg` as the url structure is generated by the cloudinary api like this:

> `https://res.cloudinary.com/<cloud_name>/<asset_type>/<delivery_type>/<transformations>/<version>/<public_id_full_path>.<extension>`

I can only influence the parts

- `<public_id_full_path>.<extension>` : if I added the width there I think I'd still get an error due to the [`urlStr === src`](https://github.com/vercel/next.js/blob/canary/packages/next/client/image.tsx#L639 ) condition.
-  `<transformations>` : but this would cost me an extra transform per image

I guess if all urls generated were returned the same as the src I would understand this warning and the directive to add the unoptimized prop, but since its only one in the set I can't quite wrap my head around it.

Unfortunately, since I have quite a few images on the page I feel like I get spammed with this warning on first load and then I have to sift through them to look for other warnings 😕

**Reply by icyJoseph** [reply]
I've tried a few things, albeit not with Cloudinary, and at some point I did see the warning.

My setup:

```tsx
// index.tsx

import type { NextPage } from "next";
import Image, { type ImageLoader } from "next/image";

const customLoader: ImageLoader = ({ src, width, quality }) => {
  const url = new URL(src);
  const pathname = url.pathname;
  const [, id] = pathname.split("/").filter(Boolean);

  return `${url.origin}/id/${id}/${width}`;
};

const Home: NextPage = () => {
  return (
    <section>
      <h1>Home</h1>

      <p>Image</p>

      <div
        style={{
          position: "relative",
          width: "50vw",
          height: "50vw",
          margin: "0 auto",
          top: "100vh",
        }}
      >
        <Image
          src="https://picsum.photos/id/237/3440" 
          alt="Labrador puppy from picsum"
          loader={customLoader}
          layout="fill"
          // priority
        />
      </div>
    </section>
  );
};

export default Home;
```

And `next.config.js`:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,

  images: {
    domains: ["picsum.photos"],
    imageSizes: [234],
    deviceSizes: [600, 1000, 3440],
    path: "https://picsum.photos/",
    loader: "custom",
  },
};

module.exports = nextConfig;
```

I can see that the `img` tag has the same src, as the one of the srcset elements. I also played around with the `priority` thinking the `warnOnce` was being tripped by LCP claims, so I placed the image under the fold instead (this is why there's a top 100vh).

At some point I did see the warning, but now it vanished...


<img width="1134" alt="Screenshot 2022-05-24 at 10 36 18" src="https://user-images.githubusercontent.com/21013447/169987954-dd057876-2d46-4827-860e-63de1bca30ab.png">

I think we'll be able to figure this out. I haven't work so much with custom loaders, but does the above somehow match your setup?

**Reply by ShuPink** [reply]
The above does match my setup! 
So sorry I should have done a minimum reproduction, I just wasn't sure how to replicate without using my cloudinary instance.

I actually managed to figure it out after you said your warning vanished, because you were right, the warning was intermittent in my situation too.  

Turns out I misinterpretted the width input from [this code here](https://github.com/vercel/next.js/blob/1cd34dd640f868b405a834e6e1755711c2cfcd33/packages/next/client/image.tsx#L629).  

I thought the `width` was from the `imageSizes` / `deviceSizes` and was being checked for every size but just warning once, but it is actually the `width` from the image (which is undefined).

```typescript 
  if (!unoptimized && loader !== defaultImageLoader) {
    const urlStr = loader({
      config,
      src,
      width: widthInt || 400, // <-- aka 400 since widthInt === undefined with layout === "fill"
      quality: qualityInt || 75,
    })
    let url: URL | undefined
    try {
      url = new URL(urlStr)
    } catch (err) {}
    if (urlStr === src || (url && url.pathname === src && !url.search)) {
      warnOnce(
        `Image with src "${src}" has a "loader" property that does not implement width. Please implement it or use the "unoptimized" property instead.` +
          `\nRead more: https://nextjs.org/docs/messages/next-image-missing-loader-width`
      )
    }
  }
```

So if the generated url from the loader is the same as the src **given a width of 400**: you'll get this warning.

My loader was written to pick an available transform that would suit a device size e.g. a 750px image for 600w `https://res.cloudinary.com/<name>/image/upload/w_750/v1/test/pic.jpg 600w`  which was why I was being warned in cases where I only had transformations smaller than 400px available (e.g. 234px), but not if I had larger (e.g. 750px).

Anyway, turns out this was a silly misinterpretation on my part... and now I know how to isolate / silence this warning.
I was considering refactoring the whole thing to use normal images too, so thanks again for taking the time to look into this, really appreciated!
